### PR TITLE
Reseeding: handle/forward failures from source RNG

### PR DESCRIPTION
**Pull request by dhardy**
Replaces #32 using code from #30. (Dependent on #34.)

Any comments @pitdicker @burdges?

The `TODO: log` part may get implemented with something like this: `#[cfg(feature="log")] debug!("Reseeding RNG")`.

### CODE REVIEW

**Comment by burdges** [comment]
If an `Rng`s is meant to be reseeded like Isaac then it might do `impl<R: Rng> Reseeder<Isaac> for R`, right?

**Comment by dhardy** [comment]
I suppose so — you're saying such an implementation could overwrite Isaac's memory directly instead of using it's public interface (e.g. `from_rng`)?

But then what we have here is more general (it can reseed any RNG).

**Comment by burdges** [comment]
I kinda meant (a) an `R: Rng` not meant to be used with `ReseedingRng` would not provide `Reseeder<R>` for any types, while (b) an `R: Rng` meant to be used with `ReseedingRng` might tend to provide `Reseeder<R>` for all `Rng`s.  If so, what does the `Reseeder<R>` type buy us?  

After some though, I suppose we actually provide `Reseeder<R>` for type level singletons that do not provide `Rng`s per se, but instead create new ones, ala `OsRng::new()`.  That makes sense.

**Comment by nixpulvis** [comment]
I'd probably want to do any logging as part of a unified separate PR.

File: src/reseeding.rs [hunk 1]
```diff
@@ -23,6 +24,9 @@ const DEFAULT_GENERATION_THRESHOLD: u64 = 32 * 1024;
 
 /// A wrapper around any RNG which reseeds the underlying RNG after it
 /// has generated a certain number of random bytes.
+/// 
+/// Note that reseeding is considered advisory only. If reseeding fails, the
+/// generator may delay reseeding or not reseed at all.
```
Comment (pitdicker) [inline src/reseeding.rs]: I think some could disagree, but 👍 from me.

File: src/reseeding.rs [hunk 2]
```diff
@@ -51,12 +55,53 @@ impl<R: Rng, Rsdr: Reseeder<R>> ReseedingRng<R, Rsdr> {
...
     pub fn reseed_if_necessary(&mut self) {
         if self.bytes_generated >= self.generation_threshold {
-            self.reseeder.reseed(&mut self.rng);
+            loop {
+                if let Err(e) = self.reseeder.reseed(&mut self.rng) {
+                    // TODO: log?
+                    if e.kind.should_wait() {
+                        // Delay reseeding
+                        let delay = max(self.generation_threshold >> 8, self.bytes_generated);
+                        self.bytes_generated -= delay;
```
Comment (pitdicker) [inline src/reseeding.rs:71]: Does it make sense to rename `bytes_generated` to something more descriptive of this delay? Maybe a countdown with `bytes_until_reseed`?

File: src/reseeding.rs [hunk 3]
```diff
@@ -51,12 +55,53 @@ impl<R: Rng, Rsdr: Reseeder<R>> ReseedingRng<R, Rsdr> {
...
+                        break;
+                    } else if e.kind.should_retry() {
+                        continue;   // immediate retry
```
Comment (pitdicker) [inline src/reseeding.rs]: Should this limit the amount of retries, to avoid an endless loop?

**Review by pitdicker — COMMENTED** [review]
The new `reseed_if_necessary` and `try_reseed_if_necessary` seem like a good solution to me!

File: src/reseeding.rs (hunk 3 above)
Comment (dhardy) [reply src/reseeding.rs]: Using a global counter like in #40 maybe?

File: src/reseeding.rs (hunk 2 above)
Comment (dhardy) [reply src/reseeding.rs:71]: It's count-up, not count-down. A more generic `bytes_count` would do, but it's less descriptive, so I'll leave as is.

**Pull request merged**